jax/core.py

```diff
@@ -1623,3 +1630,39 @@ def initial_style_staging():
       yield
     finally:
       trace_state.initial_style = prev
+
+class Float0Array:
+  """Array used to represent the tangent values of int/bool primals.
+
+  No operations are defined for this array-type by design: these values can't be
+  used in combination with non-trivial vector spaces. No implicit dtype
+  conversion is done.  To use these values, the array can be cast to float32
+  zeros using the `to_zeros` method.
+  """
+  __slots__ = ['shape']
+  dtype = dtypes.float0
+
+  __array_priority__ = 100
+
+  # Float0Array has methods that are dynamically populated in lax_numpy.py,
+  # and this annotation is needed to make pytype happy.
+  _HAS_DYNAMIC_ATTRIBUTES = True
+
+  def __init__(self, shape):
+    self.shape = shape
+
+  @property
+  def ndim(self):
+    return len(self.shape)
+
+  def to_zeros(self):
+    return np.zeros(self.shape, dtype=np.float32)
+
+  def __eq__(self, other) -> bool:
+    return type(self) is type(other) and self.shape == other.shape
```

Comment:
Nit: this definition of equality is incompatible with the definition on other JAX/NumPy arrays.